content/client-lib-development-guide/features.textile

```diff
@@ -267,7 +268,20 @@ h3(#rest-channel). Channel
 *** @(RSL6a2)@ A set of tests must exist to ensure that the client library provides interoperability for the @extras@ field which is a JSON-encodable object (ie a value that represents a JSON @object@ value and supports serialization to and from JSON text). The test, at a minimum, should publish a message with an @extras@ object such as @{"push":[{"title":"Testing"}]}@ and ensure it is received with an equivalent JSON-encodable object
-** @(RSL6b)@ If, for example, incompatible encryption details are provided or invalid Base64 is detected in the message payload, an error message will be sent to the logger, but the message will still be delivered with last successful decoding and the @encoding@ field. For example, if a message had a decoding of "utf-8/cipher+aes-128-cbc/base64", and the payload was successfully Base64 decoded but the payload could not be encrypted because the @CipherParam@ details were not configured, the message would be delivered with a binary payload and an @encoding@ with the value "utf-8/cipher+aes-128-cbc"
+** @(RSL6b)@ If, for example, incompatible encryption details are provided or invalid Base64 is detected in the message payload, an error message will be sent to the logger, but the message will still be delivered with last successful decoding and the @encoding@ field. For example, if a message had a decoding of "utf-8/cipher+aes-128-cbc/base64", and the payload was successfully Base64 decoded but the payload could not be decrypted because the @CipherParam@ details were not configured, the message would be delivered with a binary payload and an @encoding@ with the value "utf-8/cipher+aes-128-cbc. Additional steps need to be taken if decoding failed on "vcdiff" encoding, see @(RSL6c)@ 
+** @(RSL6c)@ If decoding fails on "vcdiff" encoding that means message delta application failed. Since applying a delta depends on the previous message being present a failure in the delta application on the client side makes it impossible to apply any subsequent deltas received. That would render the channel unusable so a full message needs to be requested from Ably (the server) to replace the failed delta message. That is done by:
+*** @(RSL6c1)@ Following the procedure described in "RTL4c":#RTL4c under the caveats of "RTL4d":#RTL4d, "RTL4e":#RTL4e and "RTL4f":#RTL4f. The @ProtocolMessage.channelSerial@ of the @ATTACH@ @ProtocolMessage@ sent as per "RTL4c":#RTL4c  should be set to the channel serial of the previous message to the message for which "vcdiff" decoding failed.
```

Comment:
> If decoding fails on "vcdiff" encoding that means message delta application failed.

I don't think "that means message delta application failed" is clear enough.  Perhaps:

`Given deltas are applied to the previous message published on a channel, when a "vcdiff" encoding fails to be decoded, it makes it impossible for a client to apply subsequent deltas received from that point of failure forward.  As such, the client must automatically do the following to retrieve the last complete message published:`

...

> Following the procedure described in "RTL4c":#RTL4c under the caveats of "RTL4d":#RTL4d, "RTL4e":#RTL4e and "RTL4f":#RTL4f. The @ProtocolMessage.channelSerial@ of the @ATTACH@ @ProtocolMessage@ sent as per "RTL4c":#RTL4c  should be set to the channel serial of the previous message to the message for which "vcdiff" decoding failed.

I am not clear how `RTL4d` applies in this context, the callback cannot be provided as that is provided as part of `Channel#attach`, and this error arose as part of a message decoding failure.

In regards to `RTL4e`, I don't follow why this is mentioned because a) this is unlikely, b) it's covered in full by https://docs.ably.io/client-lib-development-guide/features/#RTL14.

I think this is more complicated than it needs to be, and could perhaps be simplified as follows:

`Sending an @ATTACH@ @ProtocolMessage@ with the channel serial set to the previous message to the message for which "vcdiff" decoding failed to the server, transitioning the channel state to @ATTACHING@, and waiting for a confirmation @ATTACHED@, as per @RTL4c@ and @RTL4f@`

I have a few other questions relating to discussions we previously had:

* What happens if subsequent messages arrive, with deltas, on the channel before the `ATTACH` is sent and received, specifically in regards to emitting messages.  I understand we don't try and recovery again as per `RSL6c2`, but will we keep emitting broken messages?
* How will a user know that the message they received is broken?  AFAICT, subscribing to messages and receiving a broken delta is never going to be the end user's fault, yet will now result in:
  * A message emitted that a silent failure and a "useless" binary (unusable by the end user, and with no error)
  * Possibility that more messages being sent from the server before the `ATTACH` arrives, i.e. more broken deltas emitted
  * Duplicate messages emitted i.e. we guarantee exactly-once-delivery, yet if message 1 and 2 are received OK, 3 arrives and is broken, so the recovery process starts, and in the mean time 4 and 5 are sent, the user will receive messages 3, 4 and 5 with binary delta (effectively unusable and unexpected broken) payloads, and 3, 4 and 5 again as correct payloads.

I very much dislike everything about the silent failures, broken messages and duplicate messages, in spite of a user not doing anything wrong.

Reply:
Broken messages should not get delivered. As far as I can remember we explicitly had a discussion about duplicates/out of order messages being delivered in this scenario with @paddybyers and @SimonWoolf. It was based on [this PR comment](https://github.com/ably/ably-js/pull/615#discussion_r337599849). A decision was made that this is the behaviour we want but I am not really sure I remember all the reasoning in detail. Could you, guys, elaborate on this?

Reply:
That comment appears to be more about the channel state than what happens with messages.

Reply:
> receiving a broken delta is never going to be the end user's fault, yet will now result in: A message emitted that a silent failure and a "useless" binary (unusable by the end user, and with no error)

In what Tsviatko has implemented in JS and Java, it does not result in that; instead, if the id fails to match or it fails to vcdiff-decode, the message is discarded after logging an error, it is not delivered to the user. Eg see https://github.com/ably/ably-js/pull/615/files#diff-ee8e6cae22c16e2972f2b341a26d2ed0R451 . Further deltas on the channel will suffer the same issue, so will also be discarded.

(admittedly that is currently not made explicit in the spec, which it should be).

So in your scenario, the user will see messages 1 and 2, then a bunch of error messages to the console, then messages 3, 4, and 5, as it should be. (That's if the recovery succeeds -- if it doesn't then the `attached` will have `resumed=false` so the user will get an `update` notifying them of the discontinuity as usual).

There is one scenario where that breaks down, which is if before the reattach succeeds, the server happens to send a full message anyway, which is then processed by the client fine and emitted -- and will then be emitted again after the auto-reattach and recover. We discussed that this morning and decided that we can solve it by specifying that the client library discards messages when not in the `attached` state, and also that it goes into the `attaching` state after initiating the recovery process. That way you get no duplicates and no missing messaging.

That'd be quite a significant change, but we can't think of any reason it'd be invalid -- wdyt @mattheworiordan ?

(The other point, which @paddybyers is keen to emphasize, is that if the server correctly sticks to the protocol then these issues will never happen -- it's never 'expected'. Our (and everyone else's) guarantees are always implicitly qualified by 'assuming the server doesn't have a bug which breaks the guarantee'. So we could have decided to have just had no recovery procedure at all -- it's not like it'd be the only place where the client assumes the server is going to keep to the protocol. But here we're taking extra effort to try and keep to some subset of our guarantees even if the server has a bug (for this particular category of bugs), because we think it's a bug we're particularly likely to run into)

Reply:
> That'd be quite a significant change, but we can't think of any reason it'd be invalid -- wdyt

Agreed, can't think of an issue.